### PR TITLE
Woo/meta data refactor 2

### PR DESCRIPTION
**Pull request by hichamboushaba**
⚠️ Please don't merge this yet.

**This is the second step in the metadata refactoring.**

This PR concentrates on updating how metadata is stored in the DB, it updates the existing DB table `OrderMetaData` to be a common table to store metadata for orders and products together, this PR shouldn't have any behavioral changes, and it will be followed up with a PR that adds the logic of storing the products' metadata in the new DB table `MetaData`

#### Testing
For testing, please use the draft PR https://github.com/woocommerce/woocommerce-android/pull/12252

### CODE REVIEW

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/MetaDataDao.kt [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import androidx.room.Insert
+import androidx.room.OnConflictStrategy.Companion.REPLACE
+import androidx.room.Query
+import androidx.room.Transaction
+import kotlinx.coroutines.flow.Flow
+import org.wordpress.android.fluxc.model.LocalOrRemoteId.LocalId
+import org.wordpress.android.fluxc.persistence.entity.MetaDataEntity
+
+@Dao
+abstract class MetaDataDao {
```
Comment (hichamboushaba) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/MetaDataDao.kt:13]: I'm not sure why Github is marking the whole file as new, even though it's just a rename + some changes, and the file [history](https://github.com/wordpress-mobile/WordPress-FluxC-Android/commits/c0dd130556e99ba5cb5b3cb93e260b0a444547c9/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/MetaDataDao.kt) on Github itself shows it.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/entity/MetaDataEntity.kt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+)
+data class MetaDataEntity(
+    @ColumnInfo(name = "localSiteId")
+    val localSiteId: LocalId,
+    val id: Long,
+    val parentItemId: Long,
+    val key: String,
+    val value: String,
+    @ColumnInfo(defaultValue = "ORDER") // We default to ORDER for backwards compatibility
+    val type: ParentItemType
```
Comment (hichamboushaba) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/entity/MetaDataEntity.kt:26]: I added this for future enhancements, we might need to fetch all metadata related to some item types.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/migrations/Migrations.kt [hunk 3]
```diff
@@ -2,14 +2,16 @@ package org.wordpress.android.fluxc.persistence.migrations
...
+import androidx.room.RenameColumn
+import androidx.room.RenameTable
 import androidx.room.migration.AutoMigrationSpec
 import androidx.room.migration.Migration
 import androidx.sqlite.db.SupportSQLiteDatabase
 
 internal val MIGRATION_3_4 = object : Migration(3, 4) {
     @Suppress("LongMethod")
-    override fun migrate(database: SupportSQLiteDatabase) {
-        database.apply {
```
Comment (hichamboushaba) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/migrations/Migrations.kt:12]: All of these changes in this file are not related to the PR itself, they are just an improvement to fix a warning about the mismatch between the argument name in the overridden function and its parent.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/entity/MetaDataEntity.kt [hunk 4]
```diff
@@ -2,34 +2,24 @@ package org.wordpress.android.fluxc.persistence.entity
...
+ * The MetaDataEntity table is used to store viewable wordpress metadata. WordPress metadata
  * can potentially be quite large, so we keep it separate from the order.
+ *
+ * For now, this stores metadata for orders and products.
  */
 @Entity(
-    tableName = "OrderMetaData",
-    indices = [Index(
-        value = ["localSiteId", "orderId"]
-    )],
```
Comment (irfano) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/entity/MetaDataEntity.kt]: We removed indices here, which may cause a behavior change. Could you explain why you removed it, if this was intentional?

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/entity/MetaDataEntity.kt [hunk 5]
```diff
@@ -2,34 +2,24 @@ package org.wordpress.android.fluxc.persistence.entity
...
-    primaryKeys = ["localSiteId", "orderId", "id"],
-    foreignKeys = [ForeignKey(
-        entity = OrderEntity::class,
-        parentColumns = ["localSiteId", "orderId"],
-        childColumns = ["localSiteId", "orderId"],
-        onDelete = ForeignKey.CASCADE
-    )]
```
Comment (irfano) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/entity/MetaDataEntity.kt]: As I understand it, we will use this `MetaDataEntity` as a child table of order and product tables while it was a child table of `OrderEntity` only.
Using foreign keys was helping to provide some db issues probably. Now we removed `foreignKeys` here. Should we test this change further? Maybe we can add two foreign keys for orders and products.

**Review by irfano — COMMENTED** [review]
I'll continue my review. I have added some questions for now.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/entity/MetaDataEntity.kt (hunk 4 above)
Comment (hichamboushaba) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/entity/MetaDataEntity.kt]: Yes, it's intentional, we were just wasting resources with the index, we already use the columns as a composite primary key, which already creates an index on them.

Sqlite's documentation:
> UNIQUE and PRIMARY KEY constraints are implemented by creating a unique index in the database. 
> https://www.sqlite.org/lang_createtable.html

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/entity/MetaDataEntity.kt (hunk 5 above)
Comment (hichamboushaba) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/entity/MetaDataEntity.kt]: I thought I mentioned this as a comment, but I see now I forgot it, sorry!

> Using foreign keys was helping to provide some db issues probably

If you look at the [issue](https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/2465) that was created about this, the goal of adding foreign keys was just to allow cleaning up outdated data.

I [replaced](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/3073/commits/889d7f9935892138e3ba0251b22cce79e177ab50) this with a change to delete the metadata intentionally when we delete the orders from the DB, IMO this should be enough, as the data is saved using `ids` coming from the API, so there should be no inconsistencies, and also IMO having a common table for all of the metadata in the app will have greater benefits (will be more clear in the upcoming PRs to support Custom Fields) that outweighs the downsides of having to remove the data explicitly.

> Maybe we can add two foreign keys for orders and products

Can we? I'm not familiar with this, since the data is stored on different tables (and even worse on our case, it's on different databases).

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/entity/MetaDataEntity.kt (hunk 4 above)
Comment (irfano) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/entity/MetaDataEntity.kt]: It makes sense! Thanks for the explanation.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/entity/MetaDataEntity.kt (hunk 5 above)
Comment (irfano) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/entity/MetaDataEntity.kt]: > Can we? I'm not familiar with this, since the data is stored on different tables (and even worse on our case, it's on different databases).

Oh, I missed that `WCProductModel` uses WellSql instead of Room. In that case, it's not possible to relate `MetaData` with orders and products using foreign keys. We need to manage it manually, as you've already done in https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/3073/commits/889d7f9935892138e3ba0251b22cce79e177ab50.

**Review by irfano — APPROVED** [review]
🟢 Tested orders.
🟢 Tested updating the app from a previous version.
🟢 Tested the response of `getDisplayableOrderMetadata()`. It's the same as in the previous version, as expected.
LGTM! 👍🏻 Great job on this large task.

**Comment by hichamboushaba** [comment]
Thank you @irfano for the thourough review 🙏
